django/contrib/admin/checks.py

```diff
@@ -1184,16 +1184,21 @@ def _check_date_hierarchy(self, obj):
                     )
                 ]
             else:
-                if not isinstance(field, (models.DateField, models.DateTimeField)):
-                    return must_be(
-                        "a DateField or DateTimeField",
-                        option="date_hierarchy",
-                        obj=obj,
-                        id="admin.E128",
-                    )
-                else:
+                supported_fields = (models.DateField, models.DateTimeField)
+
+                if isinstance(field, supported_fields) or (
+                    isinstance(field, models.GeneratedField)
+                    and isinstance(field.output_field, supported_fields)
+                ):
```

Comment:
I think that checking for `output_field` like you did over `get_internal_type` is the better solution.

Reply:
Sounds good. I'll see if I can add tests/docs later today. Docs should be trivial. I'll reach out if I get stuck on tests.

Reply:
Sorry to follow up on this, as this is a release blocker we will need to have this merged by August 6th (Tuesday)
<details><summary>Here are some tests in case this helps</summary>

```diff
diff --git a/tests/modeladmin/test_checks.py b/tests/modeladmin/test_checks.py
index f767a6c92b..1206da2892 100644
--- a/tests/modeladmin/test_checks.py
+++ b/tests/modeladmin/test_checks.py
@@ -4,16 +4,17 @@ from django.contrib.admin import BooleanFieldListFilter, SimpleListFilter
 from django.contrib.admin.options import VERTICAL, ModelAdmin, TabularInline
 from django.contrib.admin.sites import AdminSite
 from django.core.checks import Error
+from django.db import models
 from django.db.models import CASCADE, F, Field, ForeignKey, ManyToManyField, Model
 from django.db.models.functions import Upper
 from django.forms.models import BaseModelFormSet
-from django.test import SimpleTestCase
+from django.test import TestCase, skipUnlessDBFeature
 from django.test.utils import isolate_apps
 
 from .models import Band, Song, User, ValidationTestInlineModel, ValidationTestModel
 
 
-class CheckTestCase(SimpleTestCase):
+class CheckTestCase(TestCase):
     def assertIsInvalid(
         self,
         model_admin,
@@ -1035,6 +1036,22 @@ class DateHierarchyCheckTests(CheckTestCase):
 
         self.assertIsValid(TestModelAdmin, ValidationTestModel)
 
+    @skipUnlessDBFeature("supports_stored_generated_columns")
+    @isolate_apps("modeladmin")
+    def test_valid_case_generated_field(self):
+        class TestModel(Model):
+            date = models.DateTimeField()
+            date_copy = models.GeneratedField(
+                expression=F("date"),
+                output_field=models.DateTimeField(),
+                db_persist=True,
+            )
+
+        class TestModelAdmin(ModelAdmin):
+            date_hierarchy = "date_copy"
+
+        self.assertIsValid(TestModelAdmin, TestModel)
+
     def test_related_valid_case(self):
         class TestModelAdmin(ModelAdmin):
             date_hierarchy = "band__sign_date"
@@ -1052,6 +1069,26 @@ class DateHierarchyCheckTests(CheckTestCase):
             "admin.E128",
         )
 
+    @skipUnlessDBFeature("supports_stored_generated_columns")
+    @isolate_apps("modeladmin")
+    def test_related_invalid_field_type_generated_field(self):
+        class TestModel(Model):
+            generated = models.GeneratedField(
+                expression=models.Value(1),
+                output_field=models.IntegerField(),
+                db_persist=True,
+            )
+
+        class TestModelAdmin(ModelAdmin):
+            date_hierarchy = "generated"
+
+        self.assertIsInvalid(
+            TestModelAdmin,
+            TestModel,
+            "The value of 'date_hierarchy' must be a DateField or DateTimeField.",
+            "admin.E128",
+        )
+
 
 class OrderingCheckTests(CheckTestCase):
```
</details>

Reply:
I believe this should work even with a 'virtual' column

Should I add tests for `supports_virtual_generated_columns` as well?

Reply:
@sarahboyce I took the example test cases you provided and fleshed them out so that it will run for both the db features `supports_stored_generated_columns` and `supports_virtual_generated_columns`, setting `db_persist` to the appropriate value.

In the future, if there's a database backend that supports BOTH features, then the test will only be run for the db_persist == False case.

If we would rather, I can create separate methods for each case and have them be thin wrappers around another assert statement.

Maybe something like this

```python
def assertGeneratedFieldIsValid(self, *, db_persist):

    class TestModel(Model):
        date = models.DateTimeField()
        date_copy = models.GeneratedField(
            expression=F("date"),
            output_field=models.DateTimeField(),
            db_persist=db_persist,
        )

    class TestModelAdmin(ModelAdmin):
        date_hierarchy = "date_copy"

    self.assertIsValid(TestModelAdmin, TestModel)

@skipUnlessDBFeature("supports_stored_generated_columns")
@isolate_apps("modeladmin")
def test_valid_case_virtual_generated_field(self):
    self.assertGeneratedFieldIsValid(db_persist=True)

@skipUnlessDBFeature("supports_virtual_generated_columns")
@isolate_apps("modeladmin")
def test_valid_case_stored_generated_field(self):
    self.assertGeneratedFieldIsValid(db_persist=False)
```

Let me know if that general approach is preferred and I'll update the pull request.

Reply:
I like the `assertGeneratedFieldIsValid` pattern :heart_eyes: 
This has been something I've found quite difficult to manage in tests in general

We might be ok just having the one though as it is enough to give us coverage for the `GeneratedField` case (Simon is more familiar with ORM test handling than me so he might have an opinion also)

Reply:
I think this version is actually more readable and technically more "correct". Having the random "assert" in the previous version was a bad smell.